fix: advertise Gateway API services via local proxy detection

vietcgi opened the pull request:
## Summary

Gateway API and Ingress services use dummy endpoints (192.192.192.192:9999) because traffic goes through Envoy, not actual pod backends. The BGP reconciler was checking `externalTrafficPolicy=Local` and wouldn't advertise when there were "no local endpoints."

**Changes:**
1. Detect Gateway/Ingress services by label (`io.cilium/gateway-api`, `io.cilium.gateway/owning-gateway`, `cilium.io/ingress`)
2. Only advertise from nodes where Envoy is running and ready
3. Verify Envoy readiness by querying the admin socket's `/ready` endpoint

This handles deployments where `cilium-envoy` has a `nodeSelector` - nodes without Envoy won't advertise Gateway routes.

Fixes #41680

## Test plan
- [x] Unit tests updated with mock `hasLocalEnvoyFunc`
- [x] Script tests use mock Envoy HTTP server on Unix socket
- [x] Existing `svc-gateway-dummy-endpoints.txtar` validates Gateway advertisement

```release-note
Fix BGP advertisement of Cilium Ingress / Gateway service's LoadBalancer IP with externalTrafficPolicy=Local
```

rastislavs commented on pkg/bgpv1/test/testdata/svc-traffic-policy.txtar:
```diff
@@ -1,7 +1,9 @@
 # Tests advertisements with external / internal service traffic policy.
-# Service routes should be withdrawn for a service with local traffic policy in case of no local endpoints.
+# LoadBalancer IPs are always advertised regardless of externalTrafficPolicy and local endpoints,
+# as they are externally-facing addresses meant for external reachability.
```
This statement is not correct, externalTrafficPolicy must be respected for LoadBalancer IPs, it was designed specifically for this. For example see https://kubernetes.io/docs/tasks/access-application-cluster/create-external-load-balancer/#preserving-the-client-source-ip

rastislavs commented on pkg/bgpv1/test/testdata/svc-gateway-dummy-endpoints.txtar:
```diff
@@ -0,0 +1,149 @@
+#! --test-peering-ips=10.99.4.101,10.99.4.102
```
Please use unique IPs for this test (each component test needs to use unique test peering IPs, so that they can run in parallel without affecting each other)

vietcgi commented:
Hi @rastislavs,

  Thanks for the detailed feedback. I've addressed all your concerns:

  IP+Port Matching: Implemented proper 192.192.192.192:9999 detection using the same approach as conversions.go
  Unique Test IPs: Changed gateway test from 10.99.4.x to 10.99.5.x for parallel execution
  Traffic Policy: Reverted svc-traffic-policy.txtar to correct expectations LoadBalancer IPs with eTP:Local respect traffic policy

rastislavs commented:
/test

vietcgi commented:
Hey @rastislavs! I fixed the checkpatch failure by rebasing instead of merging, it was complaining about the CODEOWNERS commit from main that got pulled in. Quick question about the other test failures (FQDN, EKS conformance) I traced them back and found the exact same failures on main branch (e.g., run 19303049069 has the identical FQDN nginx-ci.cilium.rocks
  failure). Since our changes only touch BGP service advertisement and don't overlap with FQDN/IPAM/egress components, should these block the merge? Or is there a process for known flakes?

rastislavs commented on pkg/datapath/linux/route/route_darwin.go:
```diff
@@ -9,15 +9,14 @@ import (
 // Replace is not supported on Darwin and will return an error at runtime.
-func Replace(route Route, mtuConfig mtu.Configuration) error {
+// The mtuConfig parameter is ignored as this is a stub implementation.
+func Replace(route Route, mtuConfig any) error {
```
Just wondering, what was the motivation behind this change? Seems unrelated to the BGP change, so probably would make sense to separate it at least to an individual commit with the clarification in its description.

rastislavs commented:
[vietcgi](https://github.com/vietcgi)

> Since our changes only touch BGP service advertisement and don't overlap with FQDN/IPAM/egress components, should these block the merge?

No, we can re-run those tests, worst case you may need to rebase your PR if they were permanently broken in main.

> Or is there a process for known flakes?

Yes, if you hit a flake that you believe is unrelated to your change, scan through the issues with the `ci/flake` label - maybe it is known, in which case it may be worth mentioning that you hit it in the existing issue. Otherwise you can open a new "Failing Test" issue.

rastislavs commented:
/test

vietcgi commented:
@rastislavs let me know if this is good.  Thanks

vietcgi commented on pkg/bgp/manager/reconciler/service.go:
```diff
@@ -613,14 +642,41 @@ func (r *ServiceReconciler) getLoadBalancerIPPaths(p ReconcileParams, svc *loadb
+		// with actual traffic handled by Envoy, not by these placeholder endpoints.
+		// Note: allBackendsDummy returns false when dummy endpoints are filtered by reflector,
+		// so we also check the service label.
+		isDummyService := allBackendsDummy(fe) || isGatewayOrIngressService
```
@rastislavs   Good catch! You're right. The allBackendsDummy check was redundant. I've removed it in the latest commit (9676d2fe4a). The label check (isGatewayOrIngressService) is sufficient since Gateway API and Ingress controllers always set their labels on the services they create. This also removes the now-unused isIngressDummyBackend function and dummy endpoint constants.

rastislavs replied:
Thanks, LGTM now

rastislavs commented:
/test

julianwiedmann reviewed (approved):
Note that if we want backports for this fix (which seems reasonable?), then a `release-note` stanza is needed in the PR description.

Do we know what releases are affected by this issue?

vietcgi commented:
@rastislavs Yes, I can do a manual backport to v1.18 after this PR is merged. I'll adapt the fix for the old pkg/bgpv1/ structure and open a separate PR targeting the v1.18 branch.

vietcgi commented:
@rastislavs I updated the fix to handle deployments where cilium-envoy has a nodeSelector that excludes some nodes.

Now we verify Envoy is actually running and ready before advertising Gateway/Ingress services by querying the admin socket's /ready endpoint. Nodes without Envoy won't advertise these routes.

rastislavs commented on pkg/bgp/manager/reconciler/service.go:
```diff
@@ -706,3 +745,39 @@ func getServicePrefixLength(fe *loadbalancer.Frontend, advert v2.BGPAdvertisemen
+		Timeout: 200 * time.Millisecond,
+	}
+
+	resp, err := client.Get("http://localhost/ready")
```
I don't think attempting an HTTP request during each BGP reconciliation is a good idea. I looked for some other approach that we could use to detect this, ideally already available in statedb - and have found it:

It's in the service statedb table: `Service.ProxyRedirect` is nil if the local proxy is not running, e.g.:

Envoy not running on the node - `proxyredirect: null`:
```
> db/get services default/cilium-ingress-basic-ingress -f yaml
name: default/cilium-ingress-basic-ingress
source: k8s
labels:
    cilium.io/ingress:
        key: cilium.io/ingress
        value: "true"
        source: k8s
annotations: {}
selector: {}
natpolicy: ""
exttrafficpolicy: Cluster
inttrafficpolicy: Cluster
forwardingmode: ""
sessionaffinity: false
sessionaffinitytimeout: 0s
loadbalancerclass: null
proxyredirect: null
healthchecknodeport: 0
loopbackhostport: false
sourceranges: []
portnames:
    http: 80
    https: 443
trafficdistribution: ""
```

Envoy is running on the node - `proxyredirect` is non-null:
```
> db/get services default/cilium-ingress-basic-ingress -f yaml
name: default/cilium-ingress-basic-ingress
source: k8s
labels:
    cilium.io/ingress:
        key: cilium.io/ingress
        value: "true"
        source: k8s
annotations: {}
selector: {}
natpolicy: ""
exttrafficpolicy: Cluster
inttrafficpolicy: Cluster
forwardingmode: ""
sessionaffinity: false
sessionaffinitytimeout: 0s
loadbalancerclass: null
proxyredirect:
    proxyport: 10770
    ports:
        - 80
healthchecknodeport: 0
loopbackhostport: false
sourceranges: []
portnames:
    http: 80
    https: 443
trafficdistribution: ""
```

julianwiedmann commented:
:wave: please have a look for how this interacts with https://github.com/cilium/cilium/pull/43558

vietcgi committed:
datapath/linux: use any type in darwin route stub

Change the mtu.Configuration parameter to `any` in the darwin stub
for Replace() to avoid import cycle issues when using the route
package from packages that don't need MTU configuration.

Signed-off-by: Kevin Vu <vietcgi@gmail.com>

vietcgi committed:
fix: advertise Gateway API services via local proxy detection

When services use proxy redirect (e.g., Gateway API or Ingress), they
should be advertised via BGP even without local backends, since traffic
is handled by the local Envoy proxy.

Signed-off-by: Kevin Vu <vietcgi@gmail.com>

vietcgi commented:
@rastislavs it's good now

rastislavs commented:
Hi @vietcgi, could you please fix these two remaining issues?
- `Error: pkg/bgp/test/script_test.go:145:31: undefined: lbmaps.HaveNetNSCookieSupport`
- the commit https://github.com/cilium/cilium/pull/42386/commits/39cd8881758ebc681c3c5823777213cb27b9e1b3 seems to be linking an unrelated issue. Could you please give it a menaingful commit description instead?

Thanks for working on this!

rastislavs commented:
/test

rastislavs commented:
/test

vietcgi commented:
@rastislavs Fixed and rebased. The test was providing HaveNetNSCookieSupport manually but reflectors.Cell already provides it. Removed the duplicate.
